test/unit/SourceLocator/Type/AutoloadSourceLocatorTest.php

```diff
@@ -302,4 +308,77 @@ public function autoload(string $className) : bool
 
         return true;
     }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testPharWorks() : void
+    {
+        require_once 'phar://' . __DIR__ . '/../../Fixture/autoload.phar/vendor/autoload.php';
+        spl_autoload_register(static function (string $class) : void {
+            if ($class !== ClassNotInPhar::class) {
+                return;
+            }
+
+            include_once __DIR__ . '/../../Fixture/ClassNotInPhar.php';
+        });
+
+        $a = new AutoloadableClassInPhar();
+        $this->assertInstanceOf(AutoloadableClassInPhar::class, $a);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testPharDoesNotWork() : void
+    {
+        require_once 'phar://' . __DIR__ . '/../../Fixture/autoload.phar/vendor/autoload.php';
+        spl_autoload_register(static function (string $class) : void {
+            if ($class !== ClassNotInPhar::class) {
+                return;
+            }
+
+            include_once __DIR__ . '/../../Fixture/ClassNotInPhar.php';
+        });
+
+        $parser            = (new BetterReflection())->phpParser();
+        $functionReflector = null;
+        $astLocator        = new AstLocator($parser, static function () use (&$functionReflector) : FunctionReflector {
+            return $functionReflector;
+        });
+        $sourceLocator     = new AutoloadSourceLocator(
+            $astLocator,
+            $parser
+        );
+        $classReflector    = new ClassReflector($sourceLocator);
+        $functionReflector = new FunctionReflector($sourceLocator, $classReflector);
+
+        $reflection = $classReflector->reflect(AutoloadableClassInPhar::class);
+
+        $this->assertSame(AutoloadableClassInPhar::class, $reflection->getName());
+    }
+
+    public function testBrokenAutoloader() : void
+    {
+        $brokenAutoloader = static function () : void {
+            throw new Exception('Autoload is broken');
+        };
+
+        spl_autoload_register($brokenAutoloader);
+
+        self::expectException(Throwable::class);
+        self::expectExceptionMessage('Autoload is broken');
+
+        try {
+            (new AutoloadSourceLocator($this->astLocator))
+                ->locateIdentifier(
+                    $this->getMockReflector(),
+                    new Identifier('Whatever', new IdentifierType(IdentifierType::IDENTIFIER_CLASS))
+                );
+
+            self::fail();
+        } finally {
+            spl_autoload_unregister($brokenAutoloader);
```

Comment:
We need to verify that the previous stream wrapper and error handler are restored at this point